src/asmcup/genetics/Gene.java

```diff
@@ -10,16 +12,27 @@ public Gene(byte[] dna, int gen, float score) {
 		this.gen = gen;
 		this.score = score;
 	}
-	
+
+	@Override
 	public int compareTo(Gene other) {
 		float d = score - other.score;
 		
-		if (d == 0) {
+		if (Math.abs(d) < 0.00000001f) {

```

Comment:
> Of course floats can be binary equivalent. Relying on it to be out of a calculation is bad.

We're not relying on anything. Score (binary) equality is merely an edge case to handle. To be 100% clear:

- We assume those floats are neither `NaN` nor `+/-Infinity`.

- Then, if the floats `a` and `b` have the same value, they are binary equivalent. Their difference is therefore precisely `+0` (there are only two zero values in IEEE floats, `-0` and `+0`, they also always behave as equal under numeric comparisons). This is the condition being checked, nothing else.

- If the floats have different values, their difference `d` _cannot be zero_. Hence, either `d < 0` xor `d > 0` is true. Therefore, `a.compareTo(b) == -(b.compareTo(a))`, which is required for the sorting algorithm in question.

> You were trying to compare if the DNA is equal

Where did I state that? All we're trying to do is introduce an ordering to a set of genes. That ordering only cares about the score (because we want to sort by it). The question "when are two genes equal?" hasn't been a concern so far.